### PR TITLE
Credit transfer device rework

### PR DESCRIPTION
**Pull request by DisturbHerb**
[OBJECTS][TGUI][CLEANLINESS][QOL][REWORK][WIKI]

## About the PR 
This PR reworks the credit transfer device from its current state. While it still facilitates direct credit transfers between accounts, a lot of the back-end code has been rewritten for cleanliness and to implement new features.

To start, it now has a TGUI interface. Players can register their card with the machine and they'll own it unless the machine is reset by themselves or by a Head of Staff.
image

The owner of the machine can set a price on the machine for a payer to transfer, who can then optionally add a percentage tip on top.
image

Once a transaction is complete, the machine will print out a receipt detailing the transaction, and the customer may elect to print themselves a copy as well.
image

On the back-end, there's also a bunch of safety stuff to ensure nothing is accidentally or intentionally altered during the payment process.

## To Do
- Get someone else who's far more handsome than I to look at the DM code because for all I know it's fugly as shit. But it works!

## Why's this needed? 
Improving the ease of use of the device with an interface and clear instructions on use, in addition to the added features, should encourage its greater use. Unfortunately, the credit transfer device is neglected by most to the point where it's just table clutter. This also enables some fun forensic accounting stuff if you're into that.

## Changelog 

```changelog
(u)DisturbHerb
(+)Credit transfer devices have been reworked and TGUI-ified.
```

### CODE REVIEW

File: tgui/packages/tgui/interfaces/Cashreg.tsx [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    <Window title={name} theme="ntos" height={240} width={300}>
+      <Window.Content>
+        <Box className="cashreg__wrapper">
+          <Button
+            className="cashreg__ownerbutton"
+            color="blue"
+            disabled={active_transaction}
+            onClick={owner ? () => act('reset') : () => act('swipe_owner')}
+            tooltip={owner ? 'Click to remove ownership' : 'Click to register ownership'}>
+            {owner ? 'Owner: ' + owner : 'Swip ID to own'}
```
Comment (ZeWaka) [inline tgui/packages/tgui/interfaces/Cashreg.tsx:35]:
```suggestion
            {owner ? 'Owner: ' + owner : 'Swipe ID to own'}
```

File: code/obj/machinery/cashreg.dm [hunk 2]
```diff
@@ -6,72 +6,269 @@ TYPEINFO(/obj/machinery/cashreg)
...
-			boutput(user, "<span class='alert'>You press the reset button, but nothing happens.</span>")
-			return
-		if (tgui_alert(user, "Reset the reader?", "Reset reader", list("Yes", "No")) == "Yes")
-			boutput(user, "<span class='alert'>Reader reset.</span>")
-			user.visible_message("<span class='alert'><B>[user]</B> resets [src].</span>")
-			mainaccount = null
+			playsound(src, 'sound/machines/printer_cargo.ogg', 50, 1)
+			SPAWN(3 SECONDS)
+				if (print_customer_copy)
+					src.print_receipt(payee, O.registered, transaction_price, transaction_tip, transaction_total, customer_copy = true)
```
Comment (ZeWaka) [inline code/obj/machinery/cashreg.dm:217]: merge skew

File: code/obj/machinery/cashreg.dm [hunk 3]
```diff
@@ -6,72 +6,269 @@ TYPEINFO(/obj/machinery/cashreg)
...
+				src.print_receipt(payee, O.registered, transaction_price, transaction_tip, transaction_total)
+
+	// Generate and create a receipt. This doesn't include the delay or the sound.
+	proc/print_receipt(payee, payer, price, tip, total, customer_copy = false)
```
Comment (ZeWaka) [inline code/obj/machinery/cashreg.dm:221]: merge skew

File: code/obj/machinery/cashreg.dm [hunk 4]
```diff
@@ -6,72 +6,269 @@ TYPEINFO(/obj/machinery/cashreg)
...
+			boutput(user, "<span class='alert'>Your account cannot currently be liquidated due to active borrows.</span>")
+			src.cancel(user)
+			return
+		if (payer_account == src.owner_account)
+			boutput(user, "<span class='alert'>You can't send funds with the owner ID to the owner ID!</span>")
+			src.cancel(user)
+			return
+
+		// Confirmation of transaction
+		var/transaction_total = src.amount + round(src.amount * src.tip) // when we get to 515 please replace this with ceil()
```
Comment (ZeWaka) [inline code/obj/machinery/cashreg.dm:188]: you do know we have ceil right now, right?

File: code/obj/machinery/cashreg.dm [hunk 5]
```diff
@@ -6,72 +6,269 @@ TYPEINFO(/obj/machinery/cashreg)
...
+	/// Safety thing to prevent other tasks from occurring during a transaction.
+	var/active_transaction = FALSE
+	/// The number of transfers the machine has processed.
+	var/transaction_count = 0
+	/// The price set by the owner of the machine.
+	var/amount = 0
+	/// The proportion of the given price to be added on as a tip, from 0 to 1.
+	var/tip = 0
+	/// Maximum amount of credits a single transaction can move.
+	var/transaction_limit = 999999
```
Comment (ZeWaka) [inline code/obj/machinery/cashreg.dm:26]:
```suggestion
	var/const/transaction_limit = 999999
```

File: code/obj/machinery/cashreg.dm [hunk 6]
```diff
@@ -6,72 +6,269 @@ TYPEINFO(/obj/machinery/cashreg)
...
 	icon_state = "scanner"
-	anchored = 1
+	req_access = list(access_heads) // Allows heads of staff to deregister owners from a cashreg.
+	anchored = TRUE
 	deconstruct_flags = DECON_SCREWDRIVER | DECON_WRENCH | DECON_MULTITOOL
-	var/datum/db_record/mainaccount = null
+	flags = TGUI_INTERACTIVE
 
+	var/datum/db_record/owner_account = null
+	var/obj/item/card/id/owner_card = null
```
Comment (ZeWaka) [inline code/obj/machinery/cashreg.dm:15]: disposing

**Review by ZeWaka — CHANGES_REQUESTED** [review]

**Comment by DisturbHerb** [comment]
Fingerprints aren't being left on the machine despite `src.add_fingerprints(user)` going off properly on `attack_hand(mob/user)` while I was stepping through the code. Will have to take a look at it later.

**Comment by BatElite** [comment]
> Fingerprints aren't being left on the machine despite `src.add_fingerprints(user)` going off properly on `attack_hand(mob/user)` while I was stepping through the code. Will have to take a look at it later.

You need FPRINT in the flags I believe

**Comment by pali6** [comment]
!merge_upstream

**Comment by pali6** [comment]
!merge_upstream

**Comment by pali6** [comment]
!merge_upstream

**Comment by pali6** [comment]
!merge_upstream

**Comment by pali6** [comment]
!merge_upstream

**Comment by pali6** [comment]
!merge_upstream

**Comment by pali6** [comment]
!merge_upstream

**Review by TobleroneSwordfish — CHANGES_REQUESTED** [review]
Stuff found while testing:
- The "click to edit price" tooltip comes up even when the reader is not owned. Also just having an arbitrary space you have to click on to edit the price isn't great.
- The "swipe ID to own" button has a tooltip that says "click to register ownership" but that only works while you're directly holding the ID.
- The "click to set tip percentage" tooltip comes up when there is no price set.
- There's no margin between the two bottom buttons, they're just smooshed together.
- Might be nice if it saved prices between each sale, since a lot of uses of it are going to be selling similarly priced items.
Otherwise this looks really cool, great for RP shops!

**Comment by TobleroneSwordfish** [comment]
As I said in discord, having the entire middle area light up and give you a tooltip even if you can't interact with it is kinda jarring, would make much more sense if it just didn't do that for non-owners.

**Review by TobleroneSwordfish — APPROVED** [review]

**Pull request merged**